File: sample-ios-objc/ViewController.m
//
//  ViewController.m
//  sample-ios-objc
//

#import "ViewController.h"
#import "CZiti-Swift.h"

@interface ViewController () <UIDocumentPickerDelegate>
@property (weak, nonatomic) IBOutlet UITextField *urlTextField;
@property (weak, nonatomic) IBOutlet UITextView *textView;
@end

@implementation ViewController

Ziti *ziti;

- (NSString*)zidFile {
    NSString *documentPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject];
    return [documentPath stringByAppendingString:@"/zid.json"];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [_urlTextField addTarget:self action:@selector(onTextFieldDidEndOnExit:) forControlEvents:UIControlEventEditingDidEndOnExit];
    
    if (ziti == NULL) {
        [self runZiti];
    }
}

- (void)runZiti {
    NSLog(@"zidFile = %@", [self zidFile]);
    ziti = [[Ziti alloc] initFromFile:[self zidFile]];
    
    if (ziti != NULL) {
        [ziti runAsync: ^(ZitiError *zErr) {
            if (zErr != NULL) {
                [self handleZitiInitError:zErr];
                return;
            }
            [ZitiUrlProtocol register:ziti :10000];
        }];
    } else {
        [self onNoIdentity];
    }
}

- (void)handleZitiInitError:(ZitiError *)zErr {
    UIAlertController* alert = [UIAlertController
                                 alertControllerWithTitle:@"Ziti Init Error"
                                 message:[zErr localizedDescription]
                                 preferredStyle:UIAlertControllerStyleAlert];
     
    
     [alert addAction:[UIAlertAction
                       actionWithTitle:@"Retry"
                       style:UIAlertActionStyleDefault
                       handler:^(UIAlertAction * action) {
         [self runZiti];
     }]];
     
     [alert addAction:[UIAlertAction
                       actionWithTitle:@"Forget This Identity"
                       style:UIAlertActionStyleDefault
                       handler:^(UIAlertAction * action) {
         
         if (ziti != NULL) {
             [ziti forget];
         }
         
         NSError *error;
         [[NSFileManager defaultManager] removeItemAtPath:[self zidFile] error:&error];
         
         [self onNoIdentity];
     }]];
    
    [alert addAction:[UIAlertAction
                      actionWithTitle:@"Exit"
                      style:UIAlertActionStyleDefault
                      handler:^(UIAlertAction * action) {
        exit(1);
    }]];
     
     dispatch_async(dispatch_get_main_queue(), ^{
         [self presentViewController:alert animated:YES completion:nil];
     });
}

- (void) onNoIdentity {
    UIAlertController* alert = [UIAlertController
                                alertControllerWithTitle:@"Ziti Identity Not Found"
                                message:@"What do you want to do now?"
                                preferredStyle:UIAlertControllerStyleAlert];
    
   
    
    [alert addAction:[UIAlertAction
                      actionWithTitle:@"Enroll"
                      style:UIAlertActionStyleDefault
                      handler:^(UIAlertAction * action) {
        [self enroll];
    }]];
    
    [alert addAction:[UIAlertAction
                      actionWithTitle:@"Exit"
                      style:UIAlertActionStyleDefault
                      handler:^(UIAlertAction * action) {
        exit(1);
    }]];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self presentViewController:alert animated:YES completion:nil];
    });
}

- (void) enroll {
    dispatch_async(dispatch_get_main_queue(), ^{
        UIDocumentPickerViewController* dp =
          [[UIDocumentPickerViewController alloc] initWithDocumentTypes:@[@"public.item"]
                                                                 inMode:UIDocumentPickerModeImport];
        dp.modalPresentationStyle = UIModalPresentationFormSheet;
        dp.allowsMultipleSelection = NO;
        dp.delegate = self;
        [self presentViewController:dp animated:YES completion:^{}];
    });
}
- (void)documentPickerWasCancelled:(UIDocumentPickerViewController *)controller {
    [self onNoIdentity];
}

- (void) documentPicker:(UIDocumentPickerViewController *)controller didPickDocumentsAtURLs:(NSArray<NSURL *> *)urls {
    
    [Ziti enroll:[urls firstObject].path : ^(ZitiIdentity *zid, ZitiError *zErr) {
        
        // Alert on Error
        if (zErr != NULL) {
            UIAlertController* alert = [UIAlertController
                                       alertControllerWithTitle:@"Enrollment Error"
                                       message:[zErr localizedDescription]
                                       preferredStyle:UIAlertControllerStyleAlert];
            [alert addAction:[UIAlertAction
                             actionWithTitle:@"OK"
                             style:UIAlertActionStyleDefault
                             handler:^(UIAlertAction * action) {
                   [self onNoIdentity];
            }]];
            [self presentViewController:alert animated:YES completion:nil];
            return;
        }
        
        // Attempt to save the identity file
        if (![zid save:[self zidFile]]) {
            UIAlertController* alert = [UIAlertController
                                           alertControllerWithTitle:@"Unable to store identity file"
                                           message:[self zidFile]
                                           preferredStyle:UIAlertControllerStyleAlert];
            [alert addAction:[UIAlertAction
                                 actionWithTitle:@"OK"
                                 style:UIAlertActionStyleDefault
                                 handler:^(UIAlertAction * action) {
                   [self onNoIdentity];
            }]];
            [self presentViewController:alert animated:YES completion:nil];
            return;
        }
        
        // All good.  Tell user "good job" and attempt to runZiti
        UIAlertController* alert = [UIAlertController
                                       alertControllerWithTitle:@"Enrolled!"
                                       message:@"You have successfully enrolled"
                                       preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction
                             actionWithTitle:@"OK"
                             style:UIAlertActionStyleDefault
                             handler:^(UIAlertAction * action) {}]];
        [self presentViewController:alert animated:YES completion:nil];
        [self runZiti];
    }];
}

- (void) onTextFieldDidEndOnExit:(id) obj {
    
    NSURL *url = [[NSURL alloc] initWithString:_urlTextField.text];
    NSURLRequest *urlReq = [[NSURLRequest alloc] initWithURL:url];
    BOOL zitiCanHandle = [ZitiUrlProtocol canInitWithRequest:urlReq];
    
    if (!zitiCanHandle) {
        UIAlertController* alert = [UIAlertController
                                   alertControllerWithTitle:@"Not a Ziti request"
                                   message:@"This request will not be routed over your Ziti nework.\nAre you sure you'd like to request this URL?"
                                   preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction
                         actionWithTitle:@"OK"
                         style:UIAlertActionStyleDefault
                         handler:^(UIAlertAction * action) {
               [self loadUrl:urlReq];
        }]];
        [self presentViewController:alert animated:YES completion:nil];
    } else {
        [self loadUrl:urlReq];
    }
    [_urlTextField resignFirstResponder];
}

-(void)setScrollableText:(NSString *)txt {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self->_textView.textStorage.mutableString setString:txt];
    });
}

- (void)loadUrl:(NSURLRequest *)urlReq {
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:urlReq
                                                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if (error != NULL) {
            [self setScrollableText:error.localizedDescription];
            return;
        }
        
        NSString *docStr = @"";
        NSHTTPURLResponse *httpResp = (NSHTTPURLResponse *)response;
        if (httpResp != NULL) {
            NSString *statusStr = [NSHTTPURLResponse localizedStringForStatusCode:httpResp.statusCode];
            docStr = [NSString stringWithFormat:@"Status: %ld (%@)\n", (long)httpResp.statusCode, statusStr];
            
            NSDictionary *dict = [httpResp allHeaderFields];
            for (NSString *key in dict) {
                docStr = [NSString stringWithFormat:@"%@%@: %@\n", docStr, key, dict[key]];
            }
            docStr = [NSString stringWithFormat:@"%@\n", docStr];
        }
        
        BOOL canPrintData = NO;
        if (data != NULL) {
            NSString *str = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            if (str != NULL) {
                canPrintData = YES;
                docStr = [NSString stringWithFormat:@"%@%@", docStr, str];
            }
        }
        
        if (!canPrintData) {
            docStr = [NSString stringWithFormat:@"%@...Unable to decode body to string...", docStr];
        }
        [self setScrollableText:docStr];
    }];
    
    [dataTask resume];
}

@end

File: cziti.sample-ios/ViewController.swift
//
//  ViewController.swift
//  cziti.sample-ios
//

import UIKit
import WebKit
import CZiti

class ViewController: UIViewController {
    @IBOutlet weak var urlTextField: UITextField!
    @IBOutlet weak var textView: UITextView!
        
    var ziti:Ziti?
    var zidFile:String {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        return paths[0].appendingPathComponent("zid.json", isDirectory: false).path
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        urlTextField.addTarget(self, action: #selector(onTextFieldDidEndOnExit), for: .editingDidEndOnExit)
        
        if ziti == nil {
            runZiti()
        }
    }
    
    func runZiti() {
        ziti = Ziti(fromFile: zidFile)
        if let ziti = ziti {
            ziti.runAsync { zErr in
                guard zErr == nil else {
                    print("Unable to run Ziti: \(String(describing: zErr!))")
                    self.handleZitiInitError(zErr!)
                    return
                }
                
                ZitiUrlProtocol.register(ziti)
            }
        } else {
            onNoIdentity()
        }
    }
    
    func enroll() {
        DispatchQueue.main.async {
            let dp = UIDocumentPickerViewController(documentTypes: ["public.item"], in: .import)
            dp.modalPresentationStyle = .formSheet
            dp.allowsMultipleSelection = false
            dp.delegate = self
            self.present(dp, animated: true, completion: nil)
        }
    }
    
    func handleZitiInitError(_ zErr:ZitiError) {
        let alert = UIAlertController(
            title:"Ziti Init Error",
            message: "\(zErr.localizedDescription)\n\nWhat do you want to do now?",
            preferredStyle: .alert)
        
        alert.addAction(UIAlertAction(
            title: NSLocalizedString("Retry", comment: "Retry"),
            style: .default,
            handler: { _ in
                self.runZiti()
        }))
        
        alert.addAction(UIAlertAction(
            title: NSLocalizedString("Forget This Identity", comment: "Forget"),
            style: .destructive,
            handler: { _ in
                if let ziti = self.ziti {
                    ziti.forget()
                }
                try? FileManager.default.removeItem(atPath: self.zidFile)
                self.onNoIdentity()
        }))
        
        alert.addAction(UIAlertAction(
            title: NSLocalizedString("Exit", comment: "Exit"),
            style: .default,
            handler: { _ in
                exit(1)
        }))
        
        DispatchQueue.main.async {
            self.present(alert, animated: true, completion: nil)
        }
    }
    
    func onNoIdentity() {
        let alert = UIAlertController(
            title:"Ziti Identity Not Found",
            message: "What do you want to do now?",
            preferredStyle: .alert)
        
        alert.addAction(UIAlertAction(
            title: NSLocalizedString("Enroll", comment: "Enroll"),
            style: .default,
            handler: { _ in
                self.enroll()
        }))
        
        alert.addAction(UIAlertAction(
            title: NSLocalizedString("Exit", comment: "Exit"),
            style: .default,
            handler: { _ in
                exit(1)
        }))
        
        DispatchQueue.main.async {
            self.present(alert, animated: true, completion: nil)
        }
    }
    
    @objc func onTextFieldDidEndOnExit(){
        if let text = urlTextField.text, let url = URL(string: text) {
            let urlReq = URLRequest(url: url)
            let zitiCanHandle = ZitiUrlProtocol.canInit(with: urlReq)
            
            if !zitiCanHandle {
                let alert = UIAlertController(
                    title:"Not a Ziti request",
                    message: "This request will not be routed over your Ziti nework.\nAre you sure you'd like to request this URL?",
                    preferredStyle: .alert)
                
                alert.addAction(UIAlertAction(
                    title: NSLocalizedString("OK", comment: "Default action"),
                    style: .default,
                    handler: { _ in
                        self.loadUrl(urlReq)
                }))
                
                alert.addAction(UIAlertAction(title: NSLocalizedString("Cancel", comment: "Cancel"), style: .cancel))
                present(alert, animated: true, completion: nil)
            } else {
                self.loadUrl(urlReq)
            }
        }
        urlTextField.resignFirstResponder()
    }
    func setScrollableText(_ txt: String) {
        DispatchQueue.main.async {
            self.textView.textStorage.mutableString.setString(txt)
        }
    }
    
    func loadUrl(_ urlReq:URLRequest) {
        URLSession.shared.dataTask(with: urlReq) { (data, response, error) in
            guard error == nil else {
                self.setScrollableText(error!.localizedDescription)
                return
            }
            
            var docStr = ""
            if let response = response as? HTTPURLResponse {
                docStr += "Status: \(response.statusCode) " +
                    "(\(HTTPURLResponse.localizedString(forStatusCode: response.statusCode)))\n" +
                    response.allHeaderFields.compactMap {
                        "\($0): \($1)\n"
                    }.joined() + "\n"
            }
            
            if let data = data, let str = String(data: data, encoding: .utf8)  {
                docStr += str
            } else {
                docStr += "...Unable to decode body to string..."
            }
            self.setScrollableText(docStr)
        }.resume()
    }
}

extension ViewController : UIDocumentPickerDelegate {
    func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        onNoIdentity()
    }
    
    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        
        Ziti.enroll(urls[0].path) { zid, zErr in
            guard let zid = zid, zErr == nil else {
                let alert = UIAlertController(
                    title:"Enrollment Error",
                    message: zErr?.localizedDescription ?? "",
                    preferredStyle: .alert)
                
                alert.addAction(UIAlertAction(
                    title: NSLocalizedString("OK", comment: "Default action"),
                    style: .default,
                    handler: { _ in
                        self.onNoIdentity()
                }))
                
                self.present(alert, animated: true, completion: nil)
                return
            }
            
            // Store zid @ zidFile
            guard zid.save(self.zidFile) else {
                let alert = UIAlertController(
                    title:"Unable to store identity file",
                    message: self.zidFile,
                    preferredStyle: .alert)
                
                alert.addAction(UIAlertAction(
                    title: NSLocalizedString("OK", comment: "Default action"),
                    style: .default,
                    handler: { _ in
                        self.onNoIdentity()
                }))
                
                self.present(alert, animated: true, completion: nil)
                return
            }
            
            let alert = UIAlertController(
                title:"Enrolled!",
                message: "You have successfully enrolled id: \(zid.id)",
                preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "Default action"), style: .default))
            self.present(alert, animated: true, completion: nil)
            self.runZiti()
        }
    }
}

extension ViewController : WKNavigationDelegate {
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        let alert = UIAlertController(
            title:"Navigation failed",
            message: error.localizedDescription,
            preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "Default action"), style: .default))
        self.present(alert, animated: true, completion: nil)
    }
    
    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        let alert = UIAlertController(
            title:"Provisional Navigation Failed",
            message: error.localizedDescription,
            preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "Default action"), style: .default))
        self.present(alert, animated: true, completion: nil)
    }
}
